docs/architecture-decisions/adr005-catalog-core-entities.md

```diff
@@ -0,0 +1,75 @@
+
+## Decision
+
+We maintain a catalog of the following core entities:
+
+* **Components** are individual pieces of software
+* **APIs** are the boundaries between different components
+* **Resources** are physical or virtual infrastructure needed to operate a component
+
+![Catalog Core Entities][catalog-core-entities]
+
+### Component
+A component is a piece of software, for example a mobile application feature, web site, backend service or data pipeline (list not exhaustive). A component can be tracked in source control, or use some existing open source or commercial software. It can implement APIs for other components to consume. In turn it might depend on APIs implemented by other components, or resources that are attached to it at runtime.
+
+Component entities are typically defined in YAML descriptor files next to the code of the component, and could look like this (actual schema will evolve):
+```yaml
+apiVersion: backstage.io/v1beta1
+kind: Component
+metadata:
+  name: my-component-name
+spec:
+  type: service
+```
+
+### API
+APIs form an abstraction that allows large software ecosystems to scale. Thus, APIs are a first class citizen in the Backstage model and the primary way to discover existing functionality in the ecosystem.
+
+APIs are implemented by components and make their boundaries explicit. They might be defined using an RPC IDL (eg in Protobuf, GraphQL or similar), a data schema (eg in Avro, TFRecord or similar), or as code interfaces (eg framework APIs in Swift, Kotlin, Java, C++, Typescript etc). In any case, APIs exposed by components need to be in a known machine-readable format so we can build further tooling and analysis on top.
+
+APIs are typically indexed from existing definitions in source control and thus wouldn't need their own descriptor files, but would be stored in the catalog somewhat like this (actual schema will evolve):
+```yaml
+apiVersion: backstage.io/v1beta1
+kind: API
+metadata:
+  name: my-component-api
+spec:
+  type: grpc
+  definition: >
+    service HelloService {
+      rpc SayHello (HelloRequest) returns (HelloResponse);
+    }
+    message HelloRequest {
+      string greeting = 1;
+    }
+    message HelloResponse {
+      string reply = 1;
+    }
+```
+
+### Resource
+Resources are the infrastructure your software needs to operate at runtime like Bigtable databases, Pub/Sub topics, S3 buckets or CDNs. Modelling them together with components and APIs will allow us to visualize and create tooling around them in Backstage.
+
+Resources are typically indexed from declarative definitions (eg Terraform, GCP Config Connector, AWS Cloud Formation) and/or inventories from cloud providers (eg GCP Asset Inventory) and thus wouldn't need their own descriptor files, but would be stored in the catalog somewhat like this (actual schema will evolve):
+```yaml
+apiVersion: backstage.io/v1beta1
+kind: Resource
+metadata:
+  name: my-component-db
+spec:
+  type: gcp-spanner
```

Comment:
Should be perhaps enable composition here to not have to invent more schemas to begin with? somewhat like replica sets in k8s:

```yaml
apiVersion: backstage.io/v1beta1
kind: Resource
metadata:
  my-component-db
spec:
  type: cnrm.cloud.google.com
  manifest:
    apiVersion: spanner.cnrm.cloud.google.com/v1beta1
    kind: SpannerInstance
    metadata:
      annotations:
        cnrm.cloud.google.com/project-id: <var>[PROJECT_ID]</var>
      labels:
        label-one: "value-one"
      name: spannerinstance-sample
    spec:
      config: regional-us-west1
      displayName: Spanner Instance Sample
      numNodes: 1
```

No idea what to do for non-kcc resources though, but I would hate to force reinvention of schemas for all the difference resources. Maybe just inline definitions similar to APIs?

Reply:
Was wondering about that too, and skimped over it for now by just having a "url" and leaving it to plugins to parse that and interact with the cloud provider APIs directly. Do you think it's worth inlining the "declared, desired state" for resources in the catalog?

Reply:
If we limit it to references I'd be fine with that too, and let other backends do the rest of the work. Just want to avoid having to invent our own schemas. I'm guessing we'd want the method for solving this to be similar to how we solve it for APIs.

Something we'll likely want to enable is the creation of frontend components that describe resources of different types in different contexts. For example a plugin that provides a Status component for KCC resources, and then being able to compose those to be able to support many different resource types. It'll probably also make sense to find common fields for any kind of resource, like status, or creation time. That'll make it a lot easier to use them in both frontend and backend plugins.

Tbh I think we should start off with the most minimal possible implementation here though, and learn from doing similar things for components before we tackle APIs and Resources. Maybe just starting out with just the type and put the rest in labels, and see where that takes us?

Reply:
Agree re: starting with the most minimal implementation. Still think we'd need an additional field for every resource though, something that allows plugins to connect the resource from the catalog with an actual cloud resource. 

As an example, let's assume we are talking about a BigTable database that is defined in a KCC manifest in a repository. On every master build, the CD system would detect the KCC manifest, create a Backstage catalog spec for it, and push that to Backstage. When the user later goes to Backstage, the BigTable plugin would see the catalog resource entries with the BigTable type and render the current state by querying the BigTable GCP API. To support this flow, we'd need two things:

1.  The CD system needs to know how to translate from the KCC manifest to the Backstage resource definition so it can ship it to Backstage.
2. The Backstage plugin needs to know how to translate from the Backstage resource definition to ids that are usable against the GCP APIs, so it can fetch the actual state of the resource

Unfortunately there doesn't seem to be a reusable URL scheme that would uniquely address any resource in GCP. Not sure about alternatives, but it'd be nice if it wasn't tied to the format it is described in (so it should work also for resources declared in Terraform, or created in the cloud console). Maybe @dflemstr could chime in here, we just hit the same problem in another place.

Reply:
@Rugvip what you said about KCC got me thinking.

We want to point to a resource somehow and relate it to a component, no matter how it came into existence. That tells me it's a ref type with a few fields (project, name etc) sufficient to resolve it on the remote side.

Then we may also want to state "oh and besides, this also seems to be KCC managed". And possibly even "and here's the complete manifest for that as far as we know at ingestion time", possibly.

But the real info, both desired and actual state, should actually be fetched from KCC itself. And that tells me that the above should just at most be auxiliary annotations, not part of the model.

Reply:
For each field we add to our models, we'll have to think, is this

- authoritative?
- expressing a relation?
- just informative?
- a best effort cache of something?

Reply:
Agree with both, this isn't a vote on push or pull when it comes to ingesting resource specs (both would work very similarly), and we probably want to keep the source of truth when it comes to desired and actual state of resources in whatever declarative infra system is used. If the "ref type" is a URL or some other combination of fields that identifies the resource can be further explored later.